xarray/backends/netCDF4_.py

```diff
@@ -485,21 +496,66 @@ def encode_variable(self, variable):
-            variable, self.format, raise_on_invalid_encoding=check_encoding
-        )
         attrs = variable.attrs.copy()
-
         fill_value = attrs.pop("_FillValue", None)
-
         encoding = _extract_nc4_variable_encoding(
             variable, raise_on_invalid=check_encoding, unlimited_dims=unlimited_dims
         )
-
+        if (
+            encoding.get("enum")
+            and attrs.get("flag_meanings")
+            and attrs.get("flag_values")
+        ):
+            var_enum_dict = {
+                k: v for k, v in zip(attrs["flag_meanings"], attrs["flag_values"])
+            }
+            enum_name = encoding.pop("enum")
+            if enum_name in self.ds.enumtypes:
+                datatype = self.ds.enumtypes[enum_name]
+                if datatype.enum_dict != var_enum_dict:
+                    error_msg = (
+                        f"Cannot save variable `{name}` because an enum"
+                        f" `{enum_name}` already exists in the Dataset but have"
+                        " a different definition. Enums are created when"
+                        " `encoding['enum']` is set by combining flag_values"
+                        " and flag_meanings attributes. To fix this error, make sure"
+                        " each variable have a unique name for `encoding['enum']` or "
+                        " if they should have the same enum, that their flag_values and"
+                        " flag_meanings are identical."
+                    )
+                    raise ValueError(error_msg)
+            else:
+                datatype = self.ds.createEnumType(
+                    variable.dtype,
+                    enum_name,
+                    var_enum_dict,
+                )
+            # Make sure the values we are trying to assign are in enums valid range.
+            error_message = (
+                "Cannot save the variable `{0}` to netCDF4: `{0}` has values, such"
+                " as `{1}`, which are not in the enum/flag_values valid values:"
+                " `{2}`. Fix the variable data or edit its flag_values and"
+                " flag_meanings attributes and try again. Note that if the enum"
+                " values are not contiguous, there might be other invalid values"
+                " too."
+            )
+            valid_values = tuple(attrs["flag_values"])
+            max_val = np.max(variable.data)
+            min_val = np.min(variable.data)
+            if max_val not in valid_values:
+                raise ValueError(error_message.format(name, max_val, valid_values))
+            if min_val not in valid_values:
+                raise ValueError(error_message.format(name, min_val, valid_values))
+            del attrs["flag_values"]
+            del attrs["flag_meanings"]
+        else:
+            datatype = _get_datatype(
+                variable, self.format, raise_on_invalid_encoding=check_encoding
+            )
+            encoding.pop("dtype", None)
```

Comment:
I was checking this. This should not be needed for this PR. It's somehow related to VLEN strings, at least it breaks one VLEN string related tests without it. I could not track down the root cause yet